text/0000-wire-reform.md

```diff
@@ -0,0 +1,244 @@
+* To extract the config value from the host object by relying on the compiler's wire metadata.
+* To signal to the adapter instance when a configured context is available by providing the new context value via `update()` API. 
+
+## Responsibilities of the wire service
+
+* To define the wire adapter protocol.
+* To provide a reference implementation of the wire adapter protocol.
+* To provide an optional abstraction for existing "legacy" wire adapter factories.
+
+## Implementation Details
+
+* `@lwc/engine` does not know about `@lwc/wire-service` and vice versa.
+* `@lwc/engine` will install a descriptor on the prototype for every wired field during the decorators registration routine.
+* `@wire` decorator will create an instance of the provided `WireAdapter` during the component initialization routine by providing the data callback as argument.
+* `@wire` decorator will invoke the adapter's `connect` and `disconnect` based on the internal hooks used per component.
+* `@wire` decorator will create a mutation tracking phase to track any access executed during the computation of the config before calling the `update` routine to be able to detect mutations on those values and issue another update on the adapter instance.
+* `@wire` decorator will detect if the adapter is expecting contextual information, and if there is a context provider defined to carry on the hand-shaking protocol between the consumer and the provider.
+* `@lwc/compiler` will provide a config function per `@wire()` declaration to produce a new config object when invoked with the `component` as the first argument. The `@wire` adapter can rely on that config function to produce a new config object at will.
+* `@lwc/wire-service` becomes Lightning Platform-specific for the most part (`register()` method) since anyone can implement the wire adapter protocol.
+
+_Backwards Compatibility Notes:_
+
+* As today, the descriptor installed on the prototype of the component by the `@wire` decorator was identical to `track` when decorating a field. This means that the component author could change the value of the field, and such change will be tracked. This is no longer the case, and even though the author can still change the value, it will not be reactive, causing no side effects on the UI of the element. This is a non-backwards compatible change, but we believe this is a very low risk for something that was not working as expected.
+
+### Wire Adapter Protocol
+
+The formalization of the wire adapter protocol is important because that enables the interoperability aspect of this feature. The adapter's code should not be aware of the component system, or even the application framework. It only cares about very specific hints to produce a stream of data. The following describes the proposed protocol:
+
+```ts
+interface WireAdapter {
+    update(config: ConfigValue, context?: ContextValue);
```

Comment:
The current implementation requires wire adapters to emit `null` when presented with incomplete `config`s. Does this new protocol guarantee that `update` is only ever invoked with fully-resolved `config`s (which implies that an incomplete `config` is an error on the component author's part)?

Reply:
This is a very good question. We have a more flexible API now. The implementation in `@lwc/wire-service` abstraction does take care of the `null` value before dispatching the config event IIRC (I will double check), but the wire protocol itself doesn't care about that anymore, that's for the adapters to decide when to go and process the new config or the new context.

Reply:
I believe the `null` handling was internal to LDS. I don't see such handling in https://github.com/salesforce/lwc/blob/master/packages/@lwc/wire-service/src/index.ts#L113-L117

Reply:
IIRC it might have been a side effect of the fact that the wire adapters were hooked up before the $ references could be expanded?

Reply:
thanks @kevinv11n, that's what I thought! we are good!